cpp/src/arrow/status.h

```diff
@@ -424,6 +426,22 @@ Status Status::operator&(Status&& s) const noexcept {
   }
 }
 
+Status Status::operator&&(const Status& s) const noexcept {
+  if (ok()) {
+    return s;
+  } else {
+    return *this;
+  }
+}
+
+Status Status::operator&&(Status&& s) const noexcept {
```

Comment:
I agree that having a non-reference parameter is preferable. We might still want two overloads, however, since we're copying in the case that `*this` is not ok:

```c++
Status Status::operator&&(Status s) const& noexcept {
  if (!ok()) return *this;
  return s;
}

Status Status::operator&&(Status s) && noexcept {
  if (!ok()) return std::move(*this);
  return s;
}
```